### PR TITLE
G4 cmp 381

### PR DESCRIPTION
**Pull request by emichaud4**
Implement a method to find rotation matrix (both valleys and anti-valleys) from miller indices instead of Euler angles. Theses changes solves some of the problems from JIRA ticket : 
-G4CMP-382 (store anti-valleys)
-G4CMP-381 (rounding errors in Euler angles methods)
-G4CMP-375 (how IV scattering must use anti-valleys instead of multiplying by -1 50% of the time) 
-G4CMP-374 (Si rotation matrices convention)
-G4CMP-338 (Ge rotation matrices convention)

Need to looks at G4CMP-375 more specifically since the convention used may affect momentum after IV scattering.

### CODE REVIEW

File: library/include/G4LatticeLogical.hh [hunk 1]
```diff
@@ -38,6 +38,7 @@
 //		precompute valley inverse transforms
 // 20200608  Fix -Wshadow warnings from tempvec
 // 20210919  M. Kelsey -- Allow SetVerboseLevel() from const instances.
+// 20231017  E. Michaud -- Add 'AddValley(const G4ThreeVector&)' to compute rotation matrix from valley's direction
```
Comment (kelseymh) [inline library/include/G4LatticeLogical.hh:41]: Please break comments at less than 80 columns.  See the lines above for examples.

File: library/src/G4LatticeLogical.cc [hunk 2]
```diff
@@ -44,6 +44,7 @@
 // 20190906  M. Kelsey -- Default IV rate model to G4CMPConfigManager value.
 // 20200520  For MT thread safety, wrap G4ThreeVector buffer in function to
 //		return thread-local instance.
+// 20231017  E. Michaud -- Add 'AddValley(const G4ThreeVector&)' to compute rotation matrix from valley's direction
```
Comment (kelseymh) [inline library/src/G4LatticeLogical.cc:47]: Please break comments at less than 80 columns. See the lines above for examples.

File: library/src/G4LatticeLogical.cc [hunk 3]
```diff
@@ -690,6 +691,148 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
   fValleyAxis.push_back(fValleyInv.back()*G4ThreeVector(1.,0.,0.));
 }
 
+// Store drifting-electron valley using valley's direction
+
+void G4LatticeLogical::AddValley(const G4ThreeVector& valleyDirVec) {
+      
+  // Find the rotation matrix elements. 
+  // The rows correspond to the rotated axis direction (first row is vx,vy,vz ; second row is a,b,c and third row is d,e,f)
```
Comment (kelseymh) [inline library/src/G4LatticeLogical.cc:699]: Please break comments at less than 80 columns.   This includes all comments in this section and anywhere else.

File: library/src/G4LatticeLogical.cc [hunk 4]
```diff
@@ -690,6 +691,148 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
   fValleyAxis.push_back(fValleyInv.back()*G4ThreeVector(1.,0.,0.));
 }
 
+// Store drifting-electron valley using valley's direction
+
+void G4LatticeLogical::AddValley(const G4ThreeVector& valleyDirVec) {
+      
+  // Find the rotation matrix elements.
```
Comment (kelseymh) [inline library/src/G4LatticeLogical.cc:698]: This description should be possible to make much shorter.

File: library/src/G4LatticeLogical.cc [hunk 5]
```diff
@@ -690,6 +691,148 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
...
+  // a = -vxvy/e/vx
+  // a = -vy/e
+    
+    
+  // Finally, we compute d and e which is the same for every case :
+  // d=-vzb
+  // e=vza
+    
+    
+  double vx=valleyDirVec.x();
```
Comment (kelseymh) [inline library/src/G4LatticeLogical.cc:756]: All of this code needs to be consolidated to make better use of G4ThreeVector and G4RotationMatrix functionality.

File: library/src/G4LatticeLogical.cc [hunk 6]
```diff
@@ -690,6 +691,148 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
...
+  vx=vx/normval;
+  vy=vy/normval;
+  vz=vz/normval;
+  double a;
+  double b;
+  double c=0; // rotated y axis must stay in the X-Y plane
+  double d;
+  double e;
+  double f;
+  double antia; // find the corresponding anti-valley rotation matrix elements
```
Comment (kelseymh) [inline library/src/G4LatticeLogical.cc:769]: You should not need to do the anti-valley stuff in parallel.  You should be able to call this function with the anti-valley direction and let it do the same work.

File: library/src/G4LatticeReader.cc [hunk 7]
```diff
@@ -32,6 +32,7 @@
 // 20180815  F. Insulla -- Added IVRateQuad
 // 20181001  M. Kelsey -- Clarify IV rate parameters systematically
 // 20190704  M. Kelsey -- Add 'ivModel' to set default IV function by material
+// 20231017  E. Michaud -- Add 'valleyDir' to set rotation matrix with valley's direction instead of euler angles
```
Comment (kelseymh) [inline library/src/G4LatticeReader.cc:35]: Please break comments at less than 80 columns.

File: library/src/G4LatticeReader.cc [hunk 8]
```diff
@@ -146,6 +147,7 @@ G4bool G4LatticeReader::ProcessToken() {
       fToken == "cij")      return ProcessStiffness();  // Elasticity element
   if (fToken == "emass")    return ProcessMassTensor();	// e- mass eigenvalues
   if (fToken == "valley")   return ProcessEulerAngles(fToken); // e- drift dirs
+  if (fToken == "valleydir")   return ProcessValleyDirection(); // miller indices valley dirs
```
Comment (kelseymh) [inline library/src/G4LatticeReader.cc:150]: Can you line up with the other lines, please?

File: library/src/G4LatticeReader.cc [hunk 9]
```diff
@@ -392,12 +394,27 @@ G4bool G4LatticeReader::ProcessEulerAngles(const G4String& name) {
     G4cerr << "G4LatticeReader: Unknown rotation matrix " << name << G4endl;
     return false;
   }
-
```
Comment (kelseymh) [inline library/src/G4LatticeReader.cc:395]: Why did you delete this line?

File: library/src/G4LatticeReader.cc [hunk 10]
```diff
@@ -392,12 +394,27 @@ G4bool G4LatticeReader::ProcessEulerAngles(const G4String& name) {
...
+
+G4bool G4LatticeReader::ProcessValleyDirection() {
+  G4double milleri=0., millerj=0., millerk=0.;
+  *psLatfile >> milleri >> millerj >> millerk;
+  if (verboseLevel>1)
+    G4cout << " ProcessValleyDirection "  << milleri << " " 
+	   << millerj << " " << millerk << G4endl;
+
+  G4ThreeVector valleyDirVec(milleri,millerj,millerk);
+  pLattice->AddValley(valleyDirVec);
```
Comment (kelseymh) [inline library/src/G4LatticeReader.cc:413]: This is where you should call `pLattice->AddValley(-valleyDirVec);` to get the anti-valley.

**Review by kelseymh — CHANGES_REQUESTED** [review]
Let me know in Slack when you've had a chance to process my comments.  If you have questions, ask in Slack or replying to these comments.  Thank you!

File: library/src/G4LatticeReader.cc [hunk 11]
```diff
@@ -392,12 +395,29 @@ G4bool G4LatticeReader::ProcessEulerAngles(const G4String& name) {
     G4cerr << "G4LatticeReader: Unknown rotation matrix " << name << G4endl;
     return false;
   }
-
+    
   G4double degOrRad = ProcessUnits("Angle");
   pLattice->AddValley(phi*degOrRad, theta*degOrRad, psi*degOrRad);
```
Comment (kelseymh) [inline library/src/G4LatticeReader.cc:400]: We need to be doing the antivalley here as well.  If the user's config.txt is using the old "valley" input, then we still need to compute the antivalley for them.  You shouldn't do it by modifying that AddValley function internally; instead, you should apply the appropriate change to the Euler angles (psi -> -psi?  Or if psi==0, something else?) and call AddValley a second time with the modified values.

**Review by kelseymh — CHANGES_REQUESTED** [review]

File: library/src/G4LatticeLogical.cc (hunk 4 above)
Comment (kelseymh) [reply library/src/G4LatticeLogical.cc:698]: I do not see any changes to the description.  The description should not simply repeat the actual calculations in the code.  It should be explaining *why* the code is written, or *how* the code does the job.  Pretend you're at a whiteboard explaining it to another student.  What words would you say to them while writing down the equations?

File: library/src/G4LatticeLogical.cc (hunk 5 above)
Comment (kelseymh) [reply library/src/G4LatticeLogical.cc]: This still needs to be done.  If you're not sure how to do it, please say so here in these comments so we can work together to resolve.  Using the vector manipulations built into G4ThreeVector and G4RotationMatrix should make the code much more self-documenting.

**Comment by kelseymh** [comment]
Please update ChangeHistory at the top with a one-line description, to make this branch ready for merging.

File: library/src/G4LatticeLogical.cc (hunk 4 above)
Comment (emichaud4) [reply library/src/G4LatticeLogical.cc:698]: I made a few changes such as shortening the explications, explaining that the matrix must be right-handed, the z axis must point upward for valleys and downward for anti-valleys, etc.). I'm not quite sure what else to change, I was trying to do something similar to ChargeCarrierTimeStep in branch G4CMP-374!

File: library/src/G4LatticeLogical.cc (hunk 5 above)
Comment (emichaud4) [reply library/src/G4LatticeLogical.cc]: I made some changes by using a tempvect(), .unit() and G4Rep3x3, I'm not sure how to consolidate that more. I like using a double for vx=tempvec().x() because we are using vx a few times in the lines below. Maybe I should change that?

File: library/src/G4LatticeReader.cc (hunk 11 above)
Comment (emichaud4) [reply library/src/G4LatticeReader.cc:400]: Perfect, I'll add the anti-valleys there!

File: library/src/G4LatticeLogical.cc (hunk 5 above)
Comment (kelseymh) [reply library/src/G4LatticeLogical.cc]: Let's find time to go through together (maybe using your slides rather than the code directly) exactly what's going on in each stage.  My suspicion is that you "should" be able to build the rotation matrix similar to how Euler angles work, by computing a set of "rotate about X", "rotate about Z" actions based on the direction vector, rather than doing the algebra "by hand" for the nine matrix components.

However, I may be falling into the "round-off trap."  That is, if we do what I'm thinking, maybe we end up with a less precise matrix than if we consolidate the underlying arithmetic the way you are doing.

File: library/src/G4LatticeLogical.cc [hunk 12]
```diff
@@ -690,6 +691,109 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
...
+   
+  else if (vx==0){ // special case if vx=0 but not vy
+      a=-1;
+      b=0;
+  }
+    
+  else if (vy==0){ // special case if vy=0 but not vx
+      a=0;
+      b=1;    
+  }
```
Comment (kelseymh) [inline library/src/G4LatticeLogical.cc:773]: if (vy == 0) { a=0; b=1; }
else if (vx == 0) { a=-1; b=0; }
else {a = -vy/f; b = vx/f; }

a = (vy==0 ? 0 : vx==0 ? -1 : -vy/f);
b = (vy==0 ? 1 : vx==0 ?  0 : vx/f);

File: library/src/G4LatticeLogical.cc [hunk 13]
```diff
@@ -690,6 +691,109 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
...
+  
+  double a;
+  double b;
+  double c=0; // rotated y axis must stay in the X-Y plane
+  double d;
+  double e;
+  double f;
+
+  // rotated z axis points in the +z direction for valleys  
+  f=sqrt(1-vz*vz);
```
Comment (kelseymh) [inline library/src/G4LatticeLogical.cc:754]: f = sqrt(1. - vz*vz) * (antival?-1:1);

File: library/src/G4LatticeLogical.cc [hunk 14]
```diff
@@ -690,6 +691,109 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
...
+  double vx=tempvec().x();
+  double vy=tempvec().y();
+  double vz=tempvec().z();
+  
+  double a;
+  double b;
+  double c=0; // rotated y axis must stay in the X-Y plane
+  double d;
+  double e;
+  double f;
```
Comment (kelseymh) [inline library/src/G4LatticeLogical.cc:751]: Don't need to declare these in advance.  Just declare the variable in the one-line calculation.

File: library/src/G4LatticeLogical.cc [hunk 15]
```diff
@@ -690,6 +691,109 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
...
+  }
+     
+  d=-vz*b;
+  e=vz*a;
+
+  G4Rep3x3 M;
+  M = G4Rep3x3( vx,vy, vz, 
+	a, b, c, 
+	d, e, f );
+
```
Comment (kelseymh) [inline library/src/G4LatticeLogical.cc:787]: Delete

File: library/src/G4LatticeLogical.cc [hunk 16]
```diff
@@ -690,6 +691,109 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
...
+	a, b, c, 
+	d, e, f );
+      
+  // Store the valley's rotation matrix, its inverse and the valley's direction.
+  fValley.resize(fValley.size()+1);
+  fValley.back().set(M);
+  fValleyInv.push_back(fValley.back());
+  fValleyInv.back().invert();
+  fValleyAxis.push_back(tempvec());
+
```
Comment (kelseymh) [inline library/src/G4LatticeLogical.cc:794]: Delete.

File: library/src/G4LatticeLogical.cc [hunk 17]
```diff
@@ -690,6 +691,109 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
...
+  d=-vz*b;
+  e=vz*a;
+
+  G4Rep3x3 M;
+  M = G4Rep3x3( vx,vy, vz, 
+	a, b, c, 
+	d, e, f );
+      
+  // Store the valley's rotation matrix, its inverse and the valley's direction.
+  fValley.resize(fValley.size()+1);
```
Comment (kelseymh) [inline library/src/G4LatticeLogical.cc:789]: fValley.back().setRows(vdir, G4ThreeVector(a,b,c), G4ThreeVector(d,e,f));

File: library/src/G4LatticeLogical.cc [hunk 18]
```diff
@@ -690,6 +691,109 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
...
+    
+  // Then we compute a (if vx=!0 and vx!=0) : 
+  // a = -bvy/vx
+  // a = -vxvy/e/vx
+  // a = -vy/e
+
+  // Finally, we compute d and e :
+  // d=-vzb
+  // e=vza 
+
```
Comment (kelseymh) [inline library/src/G4LatticeLogical.cc:737]: Delete all of this derivation.

File: library/src/G4LatticeLogical.cc [hunk 19]
```diff
@@ -690,6 +691,109 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
...
+  // First row is vx,vy,vz ; second row is a,b,c and third row is d,e,f
+    
+  // We chose the following convention :
+  //  - The rotated y axis must stay in the X-Y plane (c=0)
+  //  - Valley's rotated Z axis must have positive z component (f>0) 
+  //  - Anti-valley's rotated Z axis must have positive z component (f>0) 
+  //  - The three rotated axis should stay normalized :
+  //	vx^2 + vy^2 + vz^2 = 1 ; a^2 + b^2 = 1 ; d^2 + e^2 + f^2 = 1
+  //  - And be orthogonal :
+  //	ad + be = 0 ; avx + bvy = 0 ; dvx + evy + fvz = 0
```
Comment (kelseymh) [inline library/src/G4LatticeLogical.cc:709]: Don't need the math, just a one line statement.

File: library/src/G4LatticeLogical.cc [hunk 20]
```diff
@@ -690,6 +691,109 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
...
+  //  - It must be a right hand coordinate system :
+  //	(vx,vy,vz) X (a,b,c) = (d,e,f) : d=-vzb ; e=vza ; f=vxb - vya
```
Comment (kelseymh) [inline library/src/G4LatticeLogical.cc:711]: This is fine, but can skip showing d, e, f, expressions.  That comes from the cross-product itself.

File: library/src/G4LatticeLogical.cc [hunk 21]
```diff
@@ -690,6 +691,109 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
   fValleyAxis.push_back(fValleyInv.back()*G4ThreeVector(1.,0.,0.));
 }
 
+// Store drifting-electron valley using valley's direction
+
+void G4LatticeLogical::AddValley(const G4ThreeVector& valleyDirVec, G4bool antival) {
+      
+  // Find the rotation matrix elements. 
+  // The rows correspond to the rotated axis direction 
+  // First row is vx,vy,vz ; second row is a,b,c and third row is d,e,f
```
Comment (kelseymh) [inline library/src/G4LatticeLogical.cc:700]: Maybe do this by "drawing" the matrix (could even start on preceding line).

**Review by kelseymh — CHANGES_REQUESTED** [review]

File: ChangeHistory [hunk 22]
```diff
@@ -6,6 +6,9 @@ Releases are tagged on the 'master' branch as "g4cmp-Vxx-yy-zz".
 Features and bug fixes are tagged on the 'develop' branch using the JIRA
 ticket identifier, "G4CMP-nnn".
 
+2023-08-31  G4CMP-381 : Compute rotation matrices from valley's direction
+			instead of Euler angles. Add anti-valleys.
```
Comment (kelseymh) [inline ChangeHistory:10]: Do you think you could shorten this to one line?  If not, it's fine as it is.

File: library/src/G4LatticeReader.cc [hunk 23]
```diff
@@ -392,12 +395,31 @@ G4bool G4LatticeReader::ProcessEulerAngles(const G4String& name) {
...
+  // Anti-valleys
+  pLattice->AddValley(phi*degOrRad+pi, -1*(theta*degOrRad+pi), -psi*degOrRad);
```
Comment (kelseymh) [inline library/src/G4LatticeReader.cc:402]: Isn't the second one equivalent to just -theta*degOrRad+pi (since +pi and -pi are equivalent)?

**Review by kelseymh — CHANGES_REQUESTED** [review]

File: library/src/G4LatticeReader.cc (hunk 23 above)
Comment (emichaud4) [reply library/src/G4LatticeReader.cc]: you are right! Changed it to : -theta*degOrRad+pi

File: ChangeHistory (hunk 22 above)
Comment (emichaud4) [reply ChangeHistory]: Changed the comment to : Add valleys and anti-valleys from Miller indices.

Tell me if it's okay!
